diff --git a/src/rime/deployer.h b/src/rime/deployer.h
index c13bc10f..308e08c8 100644
--- a/src/rime/deployer.h
+++ b/src/rime/deployer.h
@@ -37,6 +37,7 @@ class Deployer : public Messenger {
   path prebuilt_data_dir;
   path staging_dir;
   path sync_dir;
+  path user_common_data_dir;
   string user_id;
   string distribution_name;
   string distribution_code_name;
diff --git a/src/rime/gear/table_translator.cc b/src/rime/gear/table_translator.cc
index 67bc2292..5491d7f5 100644
--- a/src/rime/gear/table_translator.cc
+++ b/src/rime/gear/table_translator.cc
@@ -223,6 +223,7 @@ TableTranslator::TableTranslator(const Ticket& ticket)
                     &encode_commit_history_);
     config->GetInt(name_space_ + "/max_phrase_length", &max_phrase_length_);
     config->GetInt(name_space_ + "/max_homographs", &max_homographs_);
+    config->GetInt(name_space_ + "/min_word_length", &min_word_length_);
     if (enable_sentence_ || sentence_over_completion_ ||
         contextual_suggestions_) {
       poet_.reset(new Poet(language(), config, Poet::LeftAssociateCompare));
@@ -245,6 +246,9 @@ an<Translation> TableTranslator::Query(const string& input,
                                        const Segment& segment) {
   if (!segment.HasAnyTagIn(tags_))
     return nullptr;
+  int len = static_cast<int>(input.length());
+  if (min_word_length_ > 0 && len < min_word_length_)
+    return nullptr;
   DLOG(INFO) << "input = '" << input << "', [" << segment.start << ", "
              << segment.end << ")";
 
diff --git a/src/rime/gear/table_translator.h b/src/rime/gear/table_translator.h
index 7057ed4b..78ee5d65 100644
--- a/src/rime/gear/table_translator.h
+++ b/src/rime/gear/table_translator.h
@@ -45,6 +45,7 @@ class TableTranslator : public Translator,
   bool encode_commit_history_ = true;
   int max_phrase_length_ = 5;
   int max_homographs_ = 1;
+  int min_word_length_ = 0;
   the<Poet> poet_;
   the<UnityTableEncoder> encoder_;
 };
diff --git a/src/rime/resource.cc b/src/rime/resource.cc
index 045e047f..c0de8075 100644
--- a/src/rime/resource.cc
+++ b/src/rime/resource.cc
@@ -34,10 +34,21 @@ path ResourceResolver::ResolvePath(const string& resource_id) {
 path FallbackResourceResolver::ResolvePath(const string& resource_id) {
   auto default_path = ResourceResolver::ResolvePath(resource_id);
   if (!std::filesystem::exists(default_path)) {
-    auto fallback_path = std::filesystem::absolute(
-        fallback_root_path_ / (type_.prefix + resource_id + type_.suffix));
-    if (std::filesystem::exists(fallback_path)) {
-      return fallback_path;
+    if (!fallback_root_path_.empty()) {
+      auto fallback_path = std::filesystem::absolute(
+          fallback_root_path_ /
+          (type_.prefix + resource_id + type_.suffix));
+      if (std::filesystem::exists(fallback_path)) {
+        return fallback_path;
+      }
+    }
+    if (!second_fallback_root_path_.empty()) {
+      auto second_fallback_path = std::filesystem::absolute(
+          second_fallback_root_path_ /
+          (type_.prefix + resource_id + type_.suffix));
+      if (std::filesystem::exists(second_fallback_path)) {
+        return second_fallback_path;
+      }
     }
   }
   return default_path;
diff --git a/src/rime/resource.h b/src/rime/resource.h
index 7ffa2ef4..2f1537a6 100644
--- a/src/rime/resource.h
+++ b/src/rime/resource.h
@@ -43,9 +43,13 @@ class RIME_DLL FallbackResourceResolver : public ResourceResolver {
   void set_fallback_root_path(path fallback_root_path) {
     fallback_root_path_ = fallback_root_path;
   }
+  void set_second_fallback_root_path(path second_fallback_root_path) {
+    second_fallback_root_path_ = second_fallback_root_path;
+  }
 
  private:
   path fallback_root_path_;
+  path second_fallback_root_path_;
 };
 
 }  // namespace rime
diff --git a/src/rime/service.cc b/src/rime/service.cc
index d30f991e..3e6d60d9 100644
--- a/src/rime/service.cc
+++ b/src/rime/service.cc
@@ -167,7 +167,10 @@ void Service::Notify(SessionId session_id,
 ResourceResolver* Service::CreateResourceResolver(const ResourceType& type) {
   the<FallbackResourceResolver> resolver(new FallbackResourceResolver(type));
   resolver->set_root_path(deployer().user_data_dir);
-  resolver->set_fallback_root_path(deployer().shared_data_dir);
+  if (!deployer().user_common_data_dir.empty()) {
+    resolver->set_fallback_root_path(deployer().user_common_data_dir);
+  }
+  resolver->set_second_fallback_root_path(deployer().shared_data_dir);
   return resolver.release();
 }
 
@@ -182,7 +185,11 @@ ResourceResolver* Service::CreateDeployedResourceResolver(
     const ResourceType& type) {
   the<FallbackResourceResolver> resolver(new FallbackResourceResolver(type));
   resolver->set_root_path(deployer().staging_dir);
-  resolver->set_fallback_root_path(deployer().prebuilt_data_dir);
+  if (!deployer().user_common_data_dir.empty()) {
+    resolver->set_fallback_root_path(deployer().user_common_data_dir /
+                                     "build");
+  }
+  resolver->set_second_fallback_root_path(deployer().prebuilt_data_dir);
   return resolver.release();
 }
 
diff --git a/src/rime/setup.cc b/src/rime/setup.cc
index 4e644b0e..2c52d703 100644
--- a/src/rime/setup.cc
+++ b/src/rime/setup.cc
@@ -78,6 +78,10 @@ RIME_DLL void SetupDeployer(RimeTraits* traits) {
     deployer.staging_dir = path(traits->staging_dir);
   else
     deployer.staging_dir = deployer.user_data_dir / "build";
+  const char* user_common_data_dir = getenv("RIME_USER_COMMON_DATA_DIR");
+  if (user_common_data_dir) {
+    deployer.user_common_data_dir = path(user_common_data_dir);
+  }
 }
 
 RIME_DLL void SetupLogging(const char* app_name,
diff --git a/src/rime_api.h b/src/rime_api.h
index 2fccde0f..92db29c9 100644
--- a/src/rime_api.h
+++ b/src/rime_api.h
@@ -127,6 +127,7 @@ typedef struct {
 typedef struct rime_candidate_t {
   char* text;
   char* comment;
+  char* type;
   void* reserved;
 } RimeCandidate;
 
diff --git a/src/rime_api_impl.h b/src/rime_api_impl.h
index 30f4514f..f1f8f385 100644
--- a/src/rime_api_impl.h
+++ b/src/rime_api_impl.h
@@ -203,6 +203,13 @@ static void rime_candidate_copy(RimeCandidate* dest, const an<Candidate>& src) {
   } else {
     dest->comment = nullptr;
   }
+  string type(src->type());
+  if (!type.empty()) {
+    dest->type = new char[type.length() + 1];
+    std::strcpy(dest->type, type.c_str());
+  } else {
+    dest->type = nullptr;
+  }
   dest->reserved = nullptr;
 }
 
@@ -289,6 +296,7 @@ RIME_DEPRECATED Bool RimeFreeContext(RIME_FLAVORED(RimeContext) * context) {
   for (int i = 0; i < context->menu.num_candidates; ++i) {
     delete[] context->menu.candidates[i].text;
     delete[] context->menu.candidates[i].comment;
+    delete[] context->menu.candidates[i].type;
   }
   delete[] context->menu.candidates;
   delete[] context->menu.select_keys;
@@ -404,6 +412,7 @@ RimeCandidateListNext(RimeCandidateListIterator* iterator) {
   if (auto cand = menu->GetCandidateAt((size_t)iterator->index)) {
     delete[] iterator->candidate.text;
     delete[] iterator->candidate.comment;
+    delete[] iterator->candidate.type;
     rime_candidate_copy(&iterator->candidate, cand);
     return True;
   }
@@ -415,6 +424,7 @@ RIME_DEPRECATED void RimeCandidateListEnd(RimeCandidateListIterator* iterator) {
     return;
   delete[] iterator->candidate.text;
   delete[] iterator->candidate.comment;
+  delete[] iterator->candidate.type;
   memset(iterator, 0, sizeof(RimeCandidateListIterator));
 }
 
